Support custom object types

We need to support custom object types for the next AWS release. Currently failing with:

```
panic: fatal: A failure has occurred: [pf/internal/schemashim] Failing to translate schema for attribute "custom_payload" of object type types.objectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.Message]{ObjectType:basetypes.ObjectType{AttrTypes:map[string]attr.Type{"custom_payload":types.listNestedObjectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.CustomPayload]{ListType:basetypes.ListType{ElemType:types.objectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.CustomPayload]{ObjectType:basetypes.ObjectType{AttrTypes:map[string]attr.Type{"value":basetypes.StringType{}}}}}}, "image_response_card":types.listNestedObjectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.ImageResponseCard]{ListType:basetypes.ListType{ElemType:types.objectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.ImageResponseCard]{ObjectType:basetypes.ObjectType{AttrTypes:map[string]attr.Type{"button":types.listNestedObjectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.Button]{ListType:basetypes.ListType{ElemType:types.objectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.Button]{ObjectType:basetypes.ObjectType{AttrTypes:map[string]attr.Type{"text":basetypes.StringType{}, "value":basetypes.StringType{}}}}}}, "image_url":basetypes.StringType{}, "subtitle":basetypes.StringType{}, "title":basetypes.StringType{}}}}}}, "plain_text_message":types.listNestedObjectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.PlainTextMessage]{ListType:basetypes.ListType{ElemType:types.objectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.PlainTextMessage]{ObjectType:basetypes.ObjectType{AttrTypes:map[string]attr.Type{"value":basetypes.StringType{}}}}}}, "ssml_message":types.listNestedObjectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.SSMLMessage]{ListType:basetypes.ListType{ElemType:types.objectTypeOf[github.com/hashicorp/terraform-provider-aws/internal/service/lexv2models.SSMLMessage]{ObjectType:basetypes.ObjectType{AttrTypes:map[string]attr.Type{"value":basetypes.StringType{}}}}}}}}}. This should never happen, please report an issue.
```

--- a/pf/internal/schemashim/custom_object_type_test.go
+++ b/pf/internal/schemashim/custom_object_type_test.go
@@ -1,0 +1,212 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schemashim
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+)
+
+func TestCustomObject(t *testing.T) {
+	ctx := context.Background()
+
+	obj := newObjectPseudoResource(NewObjectTypeOf[SomeType](ctx), nil, nil)
+
+	s := obj.Schema().Get("s")
+	assert.Equal(t, shim.TypeString, s.Type())
+}
+
+type SomeType struct {
+	S types.String `tfsdk:"s"`
+}
+
+// --- custom object machinery ---
+//
+// copied from terraform-provider-aws/internal/framework/types/objectof.go
+//
+// Licensed under Apache Version 2.0 HashiCorp, Inc.
+
+// objectTypeOf is the attribute type of an ObjectValueOf.
+type objectTypeOf[T any] struct {
+	basetypes.ObjectType
+}
+
+var (
+	_ basetypes.ObjectTypable  = (*objectTypeOf[struct{}])(nil)
+	_ basetypes.ObjectValuable = (*ObjectValueOf[struct{}])(nil)
+)
+
+func NewObjectTypeOf[T any](ctx context.Context) basetypes.ObjectTypable {
+	return objectTypeOf[T]{basetypes.ObjectType{AttrTypes: AttributeTypesMust[T](ctx)}}
+}
+
+func (t objectTypeOf[T]) Equal(o attr.Type) bool {
+	other, ok := o.(objectTypeOf[T])
+
+	if !ok {
+		return false
+	}
+
+	return t.ObjectType.Equal(other.ObjectType)
+}
+
+func (t objectTypeOf[T]) String() string {
+	var zero T
+	return fmt.Sprintf("ObjectTypeOf[%T]", zero)
+}
+
+func (t objectTypeOf[T]) ValueFromObject(
+	ctx context.Context, in basetypes.ObjectValue,
+) (basetypes.ObjectValuable, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	if in.IsNull() {
+		return NewObjectValueOfNull[T](ctx), diags
+	}
+	if in.IsUnknown() {
+		return NewObjectValueOfUnknown[T](ctx), diags
+	}
+
+	objectValue, d := basetypes.NewObjectValue(AttributeTypesMust[T](ctx), in.Attributes())
+	diags.Append(d...)
+	if diags.HasError() {
+		return NewObjectValueOfUnknown[T](ctx), diags
+	}
+
+	value := ObjectValueOf[T]{
+		ObjectValue: objectValue,
+	}
+
+	return value, diags
+}
+
+func (t objectTypeOf[T]) ValueFromTerraform(ctx context.Context, in tftypes.Value) (attr.Value, error) {
+	attrValue, err := t.ObjectType.ValueFromTerraform(ctx, in)
+
+	if err != nil {
+		return nil, err
+	}
+
+	objectValue, ok := attrValue.(basetypes.ObjectValue)
+
+	if !ok {
+		return nil, fmt.Errorf("unexpected value type of %T", attrValue)
+	}
+
+	objectValuable, diags := t.ValueFromObject(ctx, objectValue)
+
+	if diags.HasError() {
+		return nil, fmt.Errorf("unexpected error converting ObjectValue to ObjectValuable: %v", diags)
+	}
+
+	return objectValuable, nil
+}
+
+func (t objectTypeOf[T]) ValueType(ctx context.Context) attr.Value {
+	return ObjectValueOf[T]{}
+}
+
+// ObjectValueOf represents a Terraform Plugin Framework Object value whose corresponding Go type is the structure T.
+type ObjectValueOf[T any] struct {
+	basetypes.ObjectValue
+}
+
+func (v ObjectValueOf[T]) Equal(o attr.Value) bool {
+	other, ok := o.(ObjectValueOf[T])
+
+	if !ok {
+		return false
+	}
+
+	return v.ObjectValue.Equal(other.ObjectValue)
+}
+
+func (v ObjectValueOf[T]) Type(ctx context.Context) attr.Type {
+	return NewObjectTypeOf[T](ctx)
+}
+
+func NewObjectValueOfNull[T any](ctx context.Context) ObjectValueOf[T] {
+	return ObjectValueOf[T]{ObjectValue: basetypes.NewObjectNull(AttributeTypesMust[T](ctx))}
+}
+
+func NewObjectValueOfUnknown[T any](ctx context.Context) ObjectValueOf[T] {
+	return ObjectValueOf[T]{ObjectValue: basetypes.NewObjectUnknown(AttributeTypesMust[T](ctx))}
+}
+
+func NewObjectValueOf[T any](ctx context.Context, t *T) ObjectValueOf[T] {
+	return ObjectValueOf[T]{ObjectValue: must(basetypes.NewObjectValueFrom(ctx, AttributeTypesMust[T](ctx), t))}
+}
+
+// AttributeTypes returns a map of attribute types for the specified type T.
+// T must be a struct and reflection is used to find exported fields of T with the `tfsdk` tag.
+func AttributeTypes[T any](ctx context.Context) (map[string]attr.Type, error) {
+	var t T
+	val := reflect.ValueOf(t)
+	typ := val.Type()
+
+	if typ.Kind() == reflect.Ptr && typ.Elem().Kind() == reflect.Struct {
+		val = reflect.New(typ.Elem()).Elem()
+		typ = typ.Elem()
+	}
+
+	if typ.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("%T has unsupported type: %s", t, typ)
+	}
+
+	attributeTypes := make(map[string]attr.Type)
+	for i := 0; i < typ.NumField(); i++ {
+		field := typ.Field(i)
+		if field.PkgPath != "" {
+			continue // Skip unexported fields.
+		}
+		tag := field.Tag.Get(`tfsdk`)
+		if tag == "-" {
+			continue // Skip explicitly excluded fields.
+		}
+		if tag == "" {
+			return nil, fmt.Errorf(`%T needs a struct tag for "tfsdk" on %s`, t, field.Name)
+		}
+
+		if v, ok := val.Field(i).Interface().(attr.Value); ok {
+			attributeTypes[tag] = v.Type(ctx)
+		}
+	}
+
+	fmt.Printf("Attrs = %#v (on %T)\n", attributeTypes, t)
+
+	return attributeTypes, nil
+}
+
+func AttributeTypesMust[T any](ctx context.Context) map[string]attr.Type {
+	return must(AttributeTypes[T](ctx))
+}
+
+func must[T any, E any](t T, err E) T {
+	if v := reflect.ValueOf(err); v.IsValid() && !v.IsZero() {
+		panic(err)
+	}
+	return t
+}

--- a/pf/internal/schemashim/object_pseudoresource.go
+++ b/pf/internal/schemashim/object_pseudoresource.go
@@ -21,7 +21,6 @@ import (
 	"strconv"
 
 	"github.com/hashicorp/terraform-plugin-framework/attr"
-	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-go/tftypes"
 
 	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
@@ -108,9 +107,7 @@ func (r *objectPseudoResource) Len() int {
 
 func (r *objectPseudoResource) Get(key string) shim.Schema {
 	s, ok := r.GetOk(key)
-	if !ok {
-		panic(fmt.Sprintf("Missing key: %v", key))
-	}
+	contract.Assertf(ok, "Missing key: %v", key)
 	return s
 }
 
@@ -129,15 +126,16 @@ func (r *objectPseudoResource) GetOk(key string) (shim.Schema, bool) {
 		return newBlockSchema(key, block), true
 	}
 
-	// If there fail to find an Attr, perhaps we have a simple ObjectType then we can look up the property's
-	// AttrType and go with that.
-	if objType, ok := r.obj.(types.ObjectType); ok {
-		if t, ok := objType.AttrTypes[key]; ok {
-			return newTypeSchema(t, nil), true
-		}
+	if t, err := r.obj.ApplyTerraform5AttributePathStep(tftypes.AttributeName(key)); err == nil {
+		typ, ok := t.(attr.Type)
+		msg := "Failing to translate schema for attribute %q, unexpected attribute of type %T"
+		contract.Assertf(ok, msg, key, typ)
+		return newTypeSchema(typ, nil), true
 	}
 
+	// Check if key is a valid attribute.
 	for _, a := range r.allAttrNames {
+		// If key is a valid attribute, then we have failed to find it:
 		if key == a {
 			contract.Failf("[pf/internal/schemashim] Failing to translate schema "+
 				"for attribute %q of object type %#v. "+
@@ -146,6 +144,7 @@ func (r *objectPseudoResource) GetOk(key string) (shim.Schema, bool) {
 		}
 	}
 
+	// Otherwise key is not a valid attribute, so we can just return
 	return nil, false
 }
 
@@ -204,9 +203,7 @@ func (r *tuplePseudoResource) Schema() shim.SchemaMap {
 
 func (r *tuplePseudoResource) Get(key string) shim.Schema {
 	v, ok := r.GetOk(key)
-	if !ok {
-		panic(fmt.Sprintf("Missing key: '%s' in tuple with %d elements", key, r.Len()))
-	}
+	contract.Assertf(ok, "Missing key: '%s' in tuple with %d elements", key, r.Len())
 	return v
 }
 

--- a/pf/internal/schemashim/type_schema.go
+++ b/pf/internal/schemashim/type_schema.go
@@ -17,6 +17,7 @@ package schemashim
 import (
 	pfattr "github.com/hashicorp/terraform-plugin-framework/attr"
 	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
 	"github.com/pulumi/pulumi-terraform-bridge/pf/internal/pfutils"
 	bridge "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
 	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
@@ -57,7 +58,7 @@ func (*typeSchema) Sensitive() bool { return false }
 
 func (s *typeSchema) Elem() interface{} {
 	switch tt := s.t.(type) {
-	case types.ObjectType:
+	case basetypes.ObjectTypable:
 		var pseudoResource shim.Resource = newObjectPseudoResource(tt, s.nested, nil)
 		return pseudoResource
 	case types.ListType:

